dask/array/core.py

```diff
@@ -1573,9 +1573,479 @@ def __float__(self):
+
+            >>> parse_indices((8,), (slice(1, -1),))
+            (slice(1, 7, 1)] [])
+            >>> parse_indices((8,), ([1, 2, 4, 6],))
+            (array([1, 2, 4, 6]), [])
+            >>> parse_indices((8,), (slice(-1, 2, -1),))
+            (slice(3, 8, 1)] [0])
+            >>> parse_indices((8,), ([6, 4, 2, 1],))
+            (array([1, 2, 4, 6]), [0])
+
+            """
+            parsed_indices = []
+            mirror = []
+
+            if not isinstance(indices, tuple):
+                indices = (indices,)
+
+            # Initialize the list of parsed indices as the input
+            # indices with any Ellipsis objects expanded
+            length = len(indices)
+            n = len(shape)
+            ndim = n
+            for index in indices:
+                if index is Ellipsis:
+                    m = n - length + 1
+                    parsed_indices.extend([slice(None)] * m)
+                    n -= m
+                else:
+                    parsed_indices.append(index)
+                    n -= 1
+
+                length -= 1
+
+            len_parsed_indices = len(parsed_indices)
+
+            if ndim and len_parsed_indices > ndim:
+                raise IndexError(
+                    "Invalid indices {} for array with shape {}".format(
+                        parsed_indices, shape
+                    )
+                )
+
+            if len_parsed_indices < ndim:
+                parsed_indices.extend([slice(None)] * (ndim - len_parsed_indices))
+
+            if not ndim and parsed_indices:
+                raise IndexError("Scalar array can only be indexed with () or Ellipsis")
+
+            n_lists = 0
+
+            for i, (index, size) in enumerate(zip(parsed_indices, shape)):
+                is_slice = isinstance(index, slice)
+                if is_slice:
+                    # ------------------------------------------------
+                    # Index is a slice
+                    # ------------------------------------------------
+                    start, stop, step = index.indices(size)
+                    if step < 0 and stop == -1:
+                        stop = None
+
+                    index = slice(start, stop, step)
+
+                elif isinstance(index, (int, np.integer)):
+                    # ------------------------------------------------
+                    # Index is an integer
+                    # ------------------------------------------------
+                    if index < 0:
+                        index += size
+
+                    index = slice(index, index + 1, 1)
+                    is_slice = True
+                else:
+                    n_lists += 1
+                    if n_lists > 1:
+                        raise ValueError(
+                            "Can provide at most one dimension's index as a "
+                            "list of integers or booleans. Got: {}".format(indices)
+                        )
```

Comment:
This works in NumPy, so the error message here should be clear that this is limitation of dask's current `__setitem__` method, not a fundamental one. Also, `NotImplementedError` is probably a more appropriate type.

Reply:
Good idea.